### PR TITLE
garbage collection mechanism for AppRevision

### LINKED ISSUE #1438: [Feature] Add garbage collection mechanism for AppRevision
**Is your feature request related to a problem? Please describe.**

Currently, our apprevisions will increase Infinitely, this will cause out of memory if running long time.

**Describe the solution you'd like**

We need to add a GC mechanism for it.

By default the limitations of AppRevisions for one App could be 10. And it should be an args which could configure it in vela-core charts.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by wangyikewxgm**
fix #1438 

- [x] garbage collection mechanism for AppRevision
- [x] unit-test
- [x] e2e-test

### CODE REVIEW

File: charts/vela-core/values.yaml [hunk 1]
```diff
@@ -80,3 +80,5 @@ admissionWebhooks:
 
 
 systemDefinitionNamespace: vela-system
+
+applicationRevisionLimit: 100
```
Comment (wonderflow) [inline charts/vela-core/values.yaml:84]:
```suggestion
applicationRevisionLimit: 10
```

10 by default

File: cmd/core/main.go [hunk 2]
```diff
@@ -89,6 +89,8 @@ func main() {
 	flag.BoolVar(&logDebug, "log-debug", false, "Enable debug logs for development purpose")
 	flag.IntVar(&controllerArgs.RevisionLimit, "revision-limit", 50,
 		"RevisionLimit is the maximum number of revisions that will be maintained. The default value is 50.")
+	flag.IntVar(&controllerArgs.AppRevisionLimit, "application-revision-limit", 100,
```
Comment (wonderflow) [inline cmd/core/main.go:92]:
```suggestion
	flag.IntVar(&controllerArgs.AppRevisionLimit, "application-revision-limit", 10,
```

File: cmd/core/main.go [hunk 3]
```diff
@@ -89,6 +89,8 @@ func main() {
...
+		"application-revision-limit is the maximum number of application revisions that will be maintained. The default value is 100.")
```
Comment (wonderflow) [inline cmd/core/main.go:93]:
```suggestion
		"application-revision-limit is the maximum number of application revisions that will be maintained. The default value is 10.")
```

File: pkg/controller/core.oam.dev/oamruntime_controller.go [hunk 4]
```diff
@@ -48,6 +48,10 @@ type Args struct {
 	// The default value is 50.
 	RevisionLimit int
 
+	// AppRevisionLimit is the maximum number of application revisions that will be maintained.
+	// The default value is 100.
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/oamruntime_controller.go:52]:
```suggestion
	// The default value is 10.
```

File: pkg/controller/core.oam.dev/v1alpha2/application/revision.go [hunk 5]
```diff
@@ -291,3 +293,73 @@ func ComputeAppRevisionHash(appRevision *v1beta1.ApplicationRevision) (string, e
...
+		if err := h.r.Delete(ctx, rev.DeepCopy()); err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+		needKill--
+	}
+	return nil
+}
+
+// gatherUsingAppRevision get all using appRevisions include app's status pointing to and appContext point to
+func gatherUsingAppRevision(ctx context.Context, h *appHandler) (map[string]bool, error) {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/revision.go:336]: Do we need to check AppDelploy and AppRollout

File: pkg/controller/core.oam.dev/v1alpha2/application/revision_clean_up_test.go [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+		_, err := reconciler.Reconcile(ctrl.Request{NamespacedName: appKey})
+		Expect(err).Should(BeNil())
+		deletedRevison := new(v1beta1.ApplicationRevision)
+		revKey := types.NamespacedName{Namespace: namespace, Name: appName + "-v1"}
+		Eventually(func() error {
+			err := k8sClient.List(ctx, appRevisionList, listOpts...)
+			if err != nil {
+				return err
+			}
+			if len(appRevisionList.Items) != appRevisionLimit {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/revision_clean_up_test.go:201]: please also check the revision List contains revison v2~v6.

File: pkg/controller/core.oam.dev/v1alpha2/application/revision_clean_up_test.go [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+		checkApp.Spec.Components[0].Properties = runtime.RawExtension{Raw: []byte(property)}
+		Expect(k8sClient.Update(ctx, checkApp)).Should(BeNil())
+		_, err = reconciler.Reconcile(ctrl.Request{NamespacedName: appKey})
+		Expect(err).Should(BeNil())
+		Eventually(func() error {
+			err := k8sClient.List(ctx, appRevisionList, listOpts...)
+			if err != nil {
+				return err
+			}
+			if len(appRevisionList.Items) != appRevisionLimit {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/revision_clean_up_test.go:223]: please also check the revision List contains revison v3~v7.

File: pkg/controller/core.oam.dev/v1alpha2/application/revision.go [hunk 8]
```diff
@@ -291,3 +293,73 @@ func ComputeAppRevisionHash(appRevision *v1beta1.ApplicationRevision) (string, e
...
+	listOpts := []client.ListOption{
+		client.InNamespace(h.app.Namespace),
+		client.MatchingLabels{oam.LabelAppName: h.app.Name},
+	}
+	appRevisionList := new(v1beta1.ApplicationRevisionList)
+	// controller-runtime will cache all appRevision by default, there is no need to watch or own appRevision in manager
+	if err := h.r.List(ctx, appRevisionList, listOpts...); err != nil {
+		return err
+	}
+	needKill := len(appRevisionList.Items) - h.r.appRevisionLimit
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/revision.go:308]:
```suggestion
	needKill := len(appRevisionList.Items) - h.r.appRevisionLimit - len(usingRevision)
```

We should keep all using revisions and extra 10 ones

**Review by wonderflow — COMMENTED** [review]
good job, please fix the comments

File: pkg/controller/core.oam.dev/v1alpha2/application/revision.go (hunk 8 above)
Comment (wangyikewxgm) [reply pkg/controller/core.oam.dev/v1alpha2/application/revision.go:308]: Got it

File: pkg/controller/core.oam.dev/v1alpha2/application/apply.go [hunk 9]
```diff
@@ -587,9 +587,26 @@ func (h *appHandler) recodeTrackedResource(resourceName string, resource runtime
...
+// 1. collect useless across-namespace resource
+// 2. collect appRevision
+func garbageCollection(ctx context.Context, h *appHandler) error {
+	collectFuncs := []garbageCollectFunc{
+		garbageCollectFunc(gcAcrossNamespaceResource),
+		garbageCollectFunc(cleanUpApplicationRevision),
+	}
+	for _, collectFunc := range collectFuncs {
+		if err := collectFunc(ctx, h); err != nil {
+			return err
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/apply.go:601]: if the GC has error, log it instead of block the main logic

Comment (wonderflow) [reply pkg/controller/core.oam.dev/v1alpha2/application/apply.go:601]: after #1463 merged, we can add a event for it if GC errors

File: pkg/controller/utils/utils.go [hunk 10]
```diff
@@ -294,3 +294,45 @@ func RefreshPackageDiscover(dm discoverymapper.DiscoveryMapper, pd *definition.P
...
+		return res, nil
+	}
+	relatedRevs := new(v1beta1.ApplicationRevisionList)
+	revOpts := []client.ListOption{
+		client.InNamespace(appNs),
+		client.MatchingLabels{oam.LabelAppName: appName},
+	}
+	if err := c.List(ctx, relatedRevs, revOpts...); err != nil {
+		return nil, err
+	}
```
Comment (wonderflow) [inline pkg/controller/utils/utils.go:318]: why list it again?

File: pkg/controller/core.oam.dev/v1alpha2/application/apply.go (hunk 9 above)
Comment (wangyikewxgm) [reply pkg/controller/core.oam.dev/v1alpha2/application/apply.go:601]: Fixed

File: Makefile [hunk 11]
```diff
@@ -156,7 +156,7 @@ docker-push:
 e2e-setup:
 	helm install --create-namespace -n flux-system helm-flux http://oam.dev/catalog/helm-flux2-0.1.0.tgz
 	helm install kruise https://github.com/openkruise/kruise/releases/download/v0.7.0/kruise-chart.tgz
-	helm upgrade --install --create-namespace --namespace vela-system --set image.pullPolicy=IfNotPresent --set image.repository=vela-core-test --set image.tag=$(GIT_COMMIT) --wait kubevela ./charts/vela-core
+	helm upgrade --install --create-namespace --namespace vela-system --set image.pullPolicy=IfNotPresent --set image.repository=vela-core-test --set applicationRevisionLimit=5 --set image.tag=$(GIT_COMMIT) --wait kubevela ./charts/vela-core
```
Comment (zzxwill) [inline Makefile:159]: How about changing `appRevisioLimit` in some other files to the same as `applicationRevisionLimit`?

File: pkg/controller/core.oam.dev/v1alpha2/application/revision.go [hunk 12]
```diff
@@ -291,3 +293,81 @@ func ComputeAppRevisionHash(appRevision *v1beta1.ApplicationRevision) (string, e
...
+	needKill := len(appRevisionList.Items) - h.r.appRevisionLimit - len(usingRevision)
+	if needKill <= 0 {
+		return nil
+	}
+	h.logger.Info("application controller cleanup old appRevisions", "needKillNum", needKill)
+	sortedRevision := appRevisionList.Items
+	sort.Sort(historiesByRevision(sortedRevision))
+
+	for _, rev := range sortedRevision {
+		if needKill <= 0 {
```
Comment (zzxwill) [inline pkg/controller/core.oam.dev/v1alpha2/application/revision.go:321]:
```suggestion
		if needKill == 0 {
```

Comment (wangyikewxgm) [reply pkg/controller/core.oam.dev/v1alpha2/application/revision.go:321]: Thanks for review, i think there is no difference between them.

File: Makefile (hunk 11 above)
Comment (wangyikewxgm) [reply Makefile:159]: IMO, if user must know to use the args, we must use fullName(eg. setup-args, charts value). In code , `app` is fine : )

**Review by wonderflow — APPROVED** [review]
good job

**Review by zzxwill — APPROVED** [review]
/lgtm

**Pull request merged**